### DISCUSSION TITLE
Builders that take &mut self should panic when used a second time

### DISCUSSION THREAD

**Discussion by dtolnay** [rust-lang-nursery/api-guidelines#83] [API Guidelines]
```rust
let builder = Builder::new().a().b();
let _ = builder.build();
builder.c(); // panic
let _ = builder.build(); // panic
```

If you want to support this behavior, provide Clone for the builder.

```rust
let builder = Builder::new().a().b();
let _ = builder.clone().build();
let _ = builder.c().build();
```

In general, the user should not be allowed to rely on a builder being in any particular state after calling the terminal method (`build` or similar). If the ergonomics of doing so weren't awful, the terminal method would take self by value instead of &mut self.

**Comment by Kixunil** [comment]
Why not just consume builder instead? Compile-time checking FTW.

**Comment by dtolnay** [comment]
Because then you can't write `Builder::new().a().b().build()`.

**Comment by Kixunil** [comment]
Ah, I see. In that case I'd say consume everything.
